File: GRMCore/Class/cSetGreenAmpt.vb
Public Class cSetGreenAmpt

    Public Enum SoilTextureCode
        C
        CL
        L
        LS
        S
        SC
        SCL
        SiC
        SiCL
        SiL
        SL
        NULL
    End Enum

    Public mSoilTextureDataType As Nullable(Of cGRM.FileOrConst)
    Public mGridSoilTextureFPN As String
    Public mSoilTextureVATFPN As String
    Public mdtGreenAmptInfo As GRMProject.GreenAmptParameterDataTable
    Public mConstPorosity As Nullable(Of Single)
    Public mConstEffectivePorosity As Nullable(Of Single)
    Public mConstWFS As Nullable(Of Single)
    Public mConstHydraulicCond As Nullable(Of Single)

    Public Sub New()

    End Sub

    Public Sub SetValues(ByVal prjdb As GRMProject)
        Dim row As GRMProject.ProjectSettingsRow = CType(prjdb.ProjectSettings.Rows(0), GRMProject.ProjectSettingsRow)
        If mSoilTextureDataType.HasValue Then
            With row
                .SoilTextureDataType = mSoilTextureDataType.Value.ToString
                If mSoilTextureDataType.Value = cGRM.FileOrConst.File Then
                    .SoilTextureFile = mGridSoilTextureFPN
                    .SoilTextureVATFile = mSoilTextureVATFPN
                    .SetConstantSoilPorosityNull()
                    .SetConstantSoilEffPorosityNull()
                    .SetConstantSoilWettingFrontSuctionHeadNull()
                    .SetConstantSoilHydraulicConductivityNull()
                Else
                    .SetSoilTextureFileNull()
                    .SetSoilTextureVATFileNull()
                    .ConstantSoilPorosity = mConstPorosity.Value
                    .ConstantSoilEffPorosity = mConstEffectivePorosity.Value
                    .ConstantSoilWettingFrontSuctionHead = mConstWFS.Value
                    .ConstantSoilHydraulicConductivity = mConstHydraulicCond.Value
                End If
            End With
        End If

        If mSoilTextureDataType.Equals(cGRM.FileOrConst.File) Then
            mdtGreenAmptInfo.AcceptChanges()
            For Each r As DataRow In mdtGreenAmptInfo.Rows
                r.SetAdded()
            Next
        End If

    End Sub

    Public Sub GetValues(ByVal prjdb As GRMProject)
        mSoilTextureDataType = Nothing
        mSoilTextureVATFPN = Nothing
        mConstPorosity = Nothing
        mConstEffectivePorosity = Nothing
        mConstWFS = Nothing
        mConstHydraulicCond = Nothing
        Dim row As GRMProject.ProjectSettingsRow = CType(prjdb.ProjectSettings.Rows(0), GRMProject.ProjectSettingsRow)
        With row
            If Not .IsSoilTextureDataTypeNull Then
                If .SoilTextureDataType = cGRM.FileOrConst.File.ToString Then
                    mSoilTextureDataType = cGRM.FileOrConst.File
                    mGridSoilTextureFPN = .SoilTextureFile
                    mSoilTextureVATFPN = .SoilTextureVATFile
                ElseIf .SoilTextureDataType = cGRM.FileOrConst.Constant.ToString Then
                    mSoilTextureDataType = cGRM.FileOrConst.Constant
                    mConstPorosity = .ConstantSoilPorosity
                    mConstEffectivePorosity = .ConstantSoilEffPorosity
                    mConstWFS = .ConstantSoilWettingFrontSuctionHead
                    mConstHydraulicCond = .ConstantSoilHydraulicConductivity
                Else
                    Throw New InvalidOperationException
                End If
            End If
        End With
        If mSoilTextureDataType.Equals(cGRM.FileOrConst.File) Then
            mdtGreenAmptInfo = New GRMProject.GreenAmptParameterDataTable
            mdtGreenAmptInfo = prjdb.GreenAmptParameter
        Else
            mdtGreenAmptInfo = Nothing
        End If
    End Sub

    Public ReadOnly Property IsSet() As Boolean
        Get
            Return mSoilTextureDataType.HasValue
        End Get
    End Property

    Public Function GetSoilTextureName(ByVal intGridValue As Integer) As String
        If mSoilTextureDataType.Equals(cGRM.FileOrConst.File) Then
            Dim rows As DataRow() =
                mdtGreenAmptInfo.Select(String.Format("GridValue = {0}", intGridValue))
            Return CStr(rows(0)("GRMTextureE"))
        Else
            Return "[CONST]"
        End If
    End Function

    Public Shared Function GetSoilTextureCode(ByVal inName As String) As SoilTextureCode
        Select Case Trim(inName)
            Case SoilTextureCode.C.ToString : Return SoilTextureCode.C
            Case SoilTextureCode.CL.ToString : Return SoilTextureCode.CL
            Case SoilTextureCode.L.ToString : Return SoilTextureCode.L
            Case SoilTextureCode.LS.ToString : Return SoilTextureCode.LS
            Case SoilTextureCode.S.ToString : Return SoilTextureCode.S
            Case SoilTextureCode.SC.ToString : Return SoilTextureCode.SC
            Case SoilTextureCode.SCL.ToString : Return SoilTextureCode.SCL
            Case SoilTextureCode.SiC.ToString : Return SoilTextureCode.SiC
            Case SoilTextureCode.SiCL.ToString : Return SoilTextureCode.SiCL
            Case SoilTextureCode.SiL.ToString : Return SoilTextureCode.SiL
            Case SoilTextureCode.SL.ToString : Return SoilTextureCode.SL
            Case SoilTextureCode.NULL.ToString : Return SoilTextureCode.NULL
            Case Else : Return Nothing
        End Select
    End Function

End Class

File: GRMCore/Class/cSetGreenAmpt.cs
using System;
using System.Data;

namespace GRMCore
{
    public class cSetGreenAmpt
    {
        public enum SoilTextureCode
        {
            C,
            CL,
            L,
            LS,
            S,
            SC,
            SCL,
            SiC,
            SiCL,
            SiL,
            SL,
            USER,
            CONSTV,
            NULL
        }

        public Nullable<cGRM.FileOrConst> mSoilTextureDataType;
        public string mGridSoilTextureFPN;
        public string mSoilTextureVATFPN;
        public Dataset.GRMProject.GreenAmptParameterDataTable mdtGreenAmptInfo;
        public Nullable<double> mConstPorosity;
        public Nullable<double> mConstEffectivePorosity;
        public Nullable<double> mConstWFS;
        public Nullable<double> mConstHydraulicCond;

        public cSetGreenAmpt()
        {
        }

        public void SetValues(Dataset.GRMProject prjdb)
        {
            Dataset.GRMProject.ProjectSettingsRow row = (Dataset.GRMProject.ProjectSettingsRow)prjdb.ProjectSettings.Rows[0];
            if (mSoilTextureDataType.HasValue)
            {
                {
                    row.SoilTextureDataType = mSoilTextureDataType.Value.ToString();
                    if (mSoilTextureDataType.Value == cGRM.FileOrConst.File)
                    {
                        row.SoilTextureFile = mGridSoilTextureFPN;
                        row.SoilTextureVATFile = mSoilTextureVATFPN;
                        row.SetConstantSoilPorosityNull();
                        row.SetConstantSoilEffPorosityNull();
                        row.SetConstantSoilWettingFrontSuctionHeadNull();
                        row.SetConstantSoilHydraulicConductivityNull();
                    }
                    else
                    {
                        row.SetSoilTextureFileNull();
                        row.SetSoilTextureVATFileNull();
                        row.ConstantSoilPorosity = System.Convert.ToString(mConstPorosity.Value);
                        row.ConstantSoilEffPorosity = System.Convert.ToString(mConstEffectivePorosity.Value);
                        row.ConstantSoilWettingFrontSuctionHead = System.Convert.ToString(mConstWFS.Value);
                        row.ConstantSoilHydraulicConductivity = System.Convert.ToString(mConstHydraulicCond.Value);
                    }
                }
            }

            if (mSoilTextureDataType.Equals(cGRM.FileOrConst.File))
            {
                mdtGreenAmptInfo.AcceptChanges();
                foreach (DataRow r in mdtGreenAmptInfo.Rows)
                    r.SetAdded();
            }
        }

        public void GetValues(Dataset.GRMProject prjdb)
        {
            mSoilTextureDataType = null;
            mSoilTextureVATFPN = null;
            mConstPorosity = null;
            mConstEffectivePorosity = null;
            mConstWFS = null;
            mConstHydraulicCond = null;
            Dataset.GRMProject.ProjectSettingsRow row = (Dataset.GRMProject.ProjectSettingsRow)prjdb.ProjectSettings.Rows[0];
            if (!row.IsSoilTextureDataTypeNull())
            {
                if (row.SoilTextureDataType == cGRM.FileOrConst.File.ToString())
                {
                    mSoilTextureDataType = cGRM.FileOrConst.File;
                    mGridSoilTextureFPN = row.SoilTextureFile;
                    mSoilTextureVATFPN = row.SoilTextureVATFile;
                }
                else if (row.SoilTextureDataType == cGRM.FileOrConst.Constant.ToString())
                {
                    mSoilTextureDataType = cGRM.FileOrConst.Constant;
                    double v = 0;
                    if (double.TryParse(row.ConstantSoilPorosity, out v)) { mConstPorosity = v; }
                    if (double.TryParse(row.ConstantSoilEffPorosity, out v)) { mConstEffectivePorosity = v; }
                    if (double.TryParse(row.ConstantSoilWettingFrontSuctionHead, out v)) { mConstWFS = v; }
                    if (double.TryParse(row.ConstantSoilHydraulicConductivity, out v)) { mConstHydraulicCond = v; }
                }
                else
                { throw new InvalidOperationException(); }
            }
            if (mSoilTextureDataType.Equals(cGRM.FileOrConst.File))
            {
                mdtGreenAmptInfo = new Dataset.GRMProject.GreenAmptParameterDataTable();
                mdtGreenAmptInfo = prjdb.GreenAmptParameter;
            }
            else
                mdtGreenAmptInfo = null;
        }

        public bool IsSet
        {
            get
            {
                return mSoilTextureDataType.HasValue;
            }
        }

        public string GetSoilTextureName(int intGridValue)
        {
            if (mSoilTextureDataType.Equals(cGRM.FileOrConst.File))
            {
                DataRow[] rows = mdtGreenAmptInfo.Select(string.Format("GridValue = {0}", intGridValue));
                return rows[0]["GRMTextureE"].ToString();
            }
            else
                return "[CONST]";
        }

        public static SoilTextureCode GetSoilTextureCode(string inName)
        {
            switch (inName.Trim())
            {
                case nameof(SoilTextureCode.C):
                    {
                        return SoilTextureCode.C;
                    }

                case nameof(SoilTextureCode.CL):
                    {
                        return SoilTextureCode.CL;
                    }

                case nameof(SoilTextureCode.L):
                    {
                        return SoilTextureCode.L;
                    }

                case nameof(SoilTextureCode.LS):
                    {
                        return SoilTextureCode.LS;
                    }

                case nameof(SoilTextureCode.S):
                    {
                        return SoilTextureCode.S;
                    }

                case nameof(SoilTextureCode.SC):
                    {
                        return SoilTextureCode.SC;
                    }

                case nameof(SoilTextureCode.SCL):
                    {
                        return SoilTextureCode.SCL;
                    }

                case nameof(SoilTextureCode.SiC):
                    {
                        return SoilTextureCode.SiC;
                    }

                case nameof(SoilTextureCode.SiCL):
                    {
                        return SoilTextureCode.SiCL;
                    }

                case nameof(SoilTextureCode.SiL):
                    {
                        return SoilTextureCode.SiL;
                    }

                case nameof(SoilTextureCode.SL):
                    {
                        return SoilTextureCode.SL;
                    }

                case nameof(SoilTextureCode.USER):
                    {
                        return SoilTextureCode.USER;
                    }

                case nameof(SoilTextureCode.CONSTV):
                    {
                        return SoilTextureCode.CONSTV;
                    }

                case nameof(SoilTextureCode.NULL):
                    {
                        return SoilTextureCode.NULL;
                    }

                default:
                    {
                        return default(SoilTextureCode);
                    }
            }
        }
    }
}
